source/server/filter_chain_manager_impl.cc

```diff
@@ -215,6 +215,158 @@ void FilterChainManagerImpl::addFilterChain(
+    destination_ips.reserve(filter_chain_match.prefix_ranges().size());
+    for (const auto& destination_ip : filter_chain_match.prefix_ranges()) {
+      const auto& cidr_range = Network::Address::CidrRange::create(destination_ip);
+      destination_ips.push_back(cidr_range.asString());
+    }
+
+    std::vector<std::string> source_ips;
+    source_ips.reserve(filter_chain_match.source_prefix_ranges().size());
+    for (const auto& source_ip : filter_chain_match.source_prefix_ranges()) {
+      const auto& cidr_range = Network::Address::CidrRange::create(source_ip);
+      source_ips.push_back(cidr_range.asString());
+    }
+
+    // Reject partial wildcards, we don't match on them.
+    for (const auto& server_name : filter_chain_match.server_names()) {
+      if (server_name.find('*') != std::string::npos &&
+          !FilterChainManagerImpl::isWildcardServerName(server_name)) {
+        throw EnvoyException(
+            fmt::format("error adding listener '{}': partial wildcards are not supported in "
+                        "\"server_names\"",
+                        address_->asString()));
+      }
+    }
+
+    // Reuse created filter chain if possible.
+    // FilterChainManager maintains the lifetime of FilterChainFactoryContext
+    // ListenerImpl maintains the dependencies of FilterChainFactoryContext
+    auto filter_chain_impl = findExistingFilterChain(*filter_chain);
+    if (filter_chain_impl == nullptr) {
+      // if the build option for this filter chain is build_on_demand, we build a fake placeholder
+      // for now. Otherwise, we build the filter chain directly.
+      // TODO(ASOPVII): add build option to filter chain config.
+      bool build_on_demand = true; // filter_chain.build_option()
+      if (build_on_demand) {
+        filter_chain_impl =
+            std::make_unique<FilterChainImpl>(filter_chain); // fake filter chain placeholder
+      } else {
+        filter_chain_impl =
+            filter_chain_factory_builder.buildFilterChain(*filter_chain, context_creator);
+      }
+      ++new_filter_chain_size;
+    }
+
+    addFilterChainForDestinationPorts(
+        destination_ports_map_,
+        PROTOBUF_GET_WRAPPED_OR_DEFAULT(filter_chain_match, destination_port, 0), destination_ips,
+        filter_chain_match.server_names(), filter_chain_match.transport_protocol(),
+        filter_chain_match.application_protocols(), filter_chain_match.source_type(), source_ips,
+        filter_chain_match.source_ports(), filter_chain_impl);
+    fc_contexts_[*filter_chain] = filter_chain_impl;
+  }
+  convertIPsToTries();
+  ENVOY_LOG(debug, "new fc_contexts has {} filter chains, including {} newly built",
+            fc_contexts_.size(), new_filter_chain_size);
+}
+
+void FilterChainManagerImpl::addRealFilterChain(
+    const envoy::config::listener::v3::FilterChain* const& filter_chain,
+    FilterChainFactoryBuilder& filter_chain_factory_builder,
+    FilterChainFactoryContextCreator& context_creator) {
```

Comment:
Given one filter chain protobuf message, go request dependencies with ```FilterChainFactoryBuilder``` and ```context_creator``` and rebuild a real filter chain to replace the fake placeholder.